src/cc/BPF.cc

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cstdio>
+#include <exception>
+#include <linux/bpf.h>
+#include <regex>
+#include <unistd.h>
+#include <vector>
+
+#include "bcc_syms.h"
+#include "bpf_module.h"
+#include "libbpf.h"
+#include "perf_reader.h"
+
+#include "BPF.h"
+
+namespace ebpf {
+
+  const static std::regex KPROBE_EV_NAME_REPLACE("[+.]");
+  const static std::regex UPROBE_EV_PATH_REPLACE("[^a-zA-Z0-9_]");
+
+  BPF::~BPF() {
+    for (auto it : kprobes_)
+      if (!detach_kprobe_event(it.first, it.second))
+        fprintf(
+          stderr, "Failed to detach kprobe event %s in BPF destructor\n",
+          it.first.c_str()
+        );
+
+    for (auto it : uprobes_)
+      if (!detach_uprobe_event(it.first, it.second))
+        fprintf(
+          stderr, "Failed to detach uprobe event %s in BPF destructor\n",
+          it.first.c_str()
+        );
+    for (auto it : tracepoints_)
+      if (!detach_tracepoint_event(it.first, it.second))
+        fprintf(
+          stderr, "Failed to detach Tracepoint %s in BPF destructor\n",
+          it.first.c_str()
+        );
+    delete bpf_module_;
```

Comment:
Talked about this with @4ast and we want to avoid using `shared_ptr` for build size. It feels a bit weird if I use `unique_ptr` here but then pass the underlying raw pointer around (for tables etc.)